core/src/main/scala/kafka/log/CleanerCache.scala

```diff
@@ -0,0 +1,289 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.log
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import java.util
+
+import kafka.utils.{CoreUtils, nonthreadsafe}
+import org.apache.kafka.common.record.Record
+import org.apache.kafka.common.utils.Utils
+
+trait CleanerCache {
+  /** The maximum number of entries this map can contain */
+  def slots: Int
+
+  /**
+   * Put this record in the cache, but only if it is greater than
+   * the currently associated record (if any).
```

Comment:
Please document how the Record's are compared when choosing the greater one.
You can refer to #greater() below

Reply:
Updated the description to include some details on what 'greater' means.
Sadly, linking here is not that nice, so I've put the text in directly, I hope that is ok.